### DISCUSSION TITLE
How would I implement a random jitter for randomly scheduled jobs?

### DISCUSSION THREAD

**Discussion by hans-notifi** [quartznet/quartznet#2227] [Q&A] [answered]
I have a job that I'm creating in my ASP.NET service that runs some internal archiving task:

```
var builder = WebApplication.CreateBuilder(args);
// ... more boilerplate ...
builder.Services.AddQuartz(q =>
{
    var archiveJobKey = new JobKey("MyArchivingJob");
    q.AddJob<MyArchivingJob>(o => o.WithIdentity(archiveJobKey));
    q.AddTrigger(o =>
        o.ForJob(archiveJobKey)
            .WithIdentity("MyArchivingJob-Trigger")
            .WithSimpleSchedule(s => s.WithInterval(TimeSpan.FromHours(1)).RepeatForever()));
});
builder.Services.AddQuartzHostedService(o =>
{
    o.WaitForJobsToComplete = true;
});
builder.Services.AddTransient<MyArchivingJob>();
```

I would like to be able to run this job with a random jitter in the schedule, perhaps as wide as 10 minutes around the regular interval. This is because I have multiple copies of this service running and would like to ensure the job gets done, but do not want all of the copies running the job at the same time. I didn't notice a capability for this on any of the default schedules (`SimpleSchedule`, `CronSchedule`, `DailyTimeIntervalSchedule`, or `CalendarInterval`) - is there a way to enable it, a simple way to implement it myself, or a third-party schedule option that I could use?

**Comment by lahma** [comment, marked as answer]
For random jitter it's easiest to add a sleep (Task.Delay) in start of your job. This of course means that `max_jitter + max_job_execution_time < trigger_interval` must be true so that overlapping/queuing won't happen.

But your real problem might be that you should run database-backed job store that ensures that multiple instances of job are handled correctly - you can add `DisallowConcurrentExecution` attribute and cluster will obey that. Remember that concurrent execution checks are based on job keys.

**Comment by GFoley83** [comment]
I've been using the following extension method, in conjunction with `[DisallowConcurrentExecution]`, to implement a random delay/jitter between the current and next execution. It has been working well for me:

Usage:
```cs
    public async Task Execute(IJobExecutionContext context)
    {
        await context.AddRandomDelayBetweenExecutions();
        . . . . 
    }
```

```cs
public static class Extensions
{
    private static readonly Random _random = new();

    public static async Task<IJobExecutionContext> AddRandomDelayBetweenExecutions(this IJobExecutionContext context)
    {
        if (!context.NextFireTimeUtc.HasValue)
        {
            return context;
        }

        var randomDelay = GetRandomMilliSeconds(context.FireTimeUtc.DateTime, context.NextFireTimeUtc.Value.DateTime);

        await Task.Delay(randomDelay);

        return context;
    }

    public static int GetRandomMilliSeconds(DateTime start, DateTime end)
    {
        // Calculate the total difference in seconds
        var timeSpan = end - start;
        var totalMilliseconds = (int) Math.Abs(timeSpan.TotalMilliseconds);

        if (totalMilliseconds == 0)
        {
            return 0; // No difference
        }

        // Generate a random number between 0 and totalSeconds
        var randomMilliSeconds = _random.Next(0, totalMilliseconds + 1);
        return randomMilliSeconds;
    }
}
```
